start setup and structure

rotegras opened the pull request:
added local googlefonts.
added normalize.css.
added router structure.
added global css styles.
added folder structure.

rotegras committed:
added local googlefonts.
added normalize.css.
added router structure.
added global css styles.
added folder structure.

jkettmann commented on src/App.css:
```diff
@@ -0,0 +1,66 @@
+  font-family: 'Montserrat';
+  font-style: normal;
+  font-weight: 400;
+  src: local('Montserrat Regular'), local('Montserrat-Regular'),
```
It's easier when you add the fonts via a link in index.html. You can build this link on https://fonts.google.com/ when you select the required fonts. You can customize the required font-weights as well. Or you simply use this ;-) 

<link href="https://fonts.googleapis.com/css?family=Bitter|Montserrat:400,500,600&display=swap" rel="stylesheet">

rotegras replied:
I started self-hosting googlefonts due to gdpr. In my last job they did this way.

jkettmann replied:
Ok, that may be a valid reaso :-)

jkettmann commented on src/App.css:
```diff
@@ -0,0 +1,66 @@
+@import-normalize; /* bring in normalize.css styles */
+
+/* Fonts */
+@font-face {
```
Global font definitions :+1:

jkettmann commented on src/App.css:
```diff
@@ -0,0 +1,66 @@
+  height: 100%;
+}
+
+.page {
```
This file should be split in two. One called something like `global.css` and the other `App.css`. The `page` class is the only one specific to App.js

global.css can be imported in index.js

jkettmann commented on src/App.js:
```diff
@@ -1,10 +1,37 @@
+import Header from './components/Header/Header';
+import Search from './pages/Search';
+import HowItWorks from './pages/HowItWorks';
+import './App.css';
```
I would suggest to either go for `css-modules` or CSS-in-JS. I personally like `styled-components` the most, but `css-modules` is totally fine as well. It works out-of-the-box with create-react-app. The only thing you have to do here is change your import and usage of classes

```
import styles from './App.css';

<div className={styles.page}>
```

The benefit is that the class names are not global so you don't run into conflicts. There are tools in place that automatically namespace the classes to the component that's used. You can check it in the browser's dev tools

rotegras replied:
Started playing with styled components. It will be more interesting when I get props to pass in. But then I should probably switch completely to this system? I will get there.

jkettmann commented on src/components/Header/Header.css:
```diff
@@ -0,0 +1,34 @@
+header {
```
create a class instead of targeting a tag. So rather .header instead of header here

jkettmann commented on src/components/Header/Header.css:
```diff
@@ -0,0 +1,34 @@
+header {
+  height: 80px;
```
I guess it should be 100px per design

jkettmann commented on src/components/Header/Header.css:
```diff
@@ -0,0 +1,34 @@
+  padding: 1rem 2rem;
+}
+
+.logo a {
```
It's better to create a class for the link directly. You can call it logoLink or so. The following two rules can be combined

```
.logoLink {
  font-size: 1rem;
  text-decoration: none;

  :focus {
    outline: none;
  }
}
```

The reason why it's better to have specific classes for each element instead of nesting rules like `.logo a` or `.something div` is this:

Let's say you have a div inside an li and you target it like `.listItem div { ... }`. Now you're styles are highly coupled to the DOM structure. When you change the inner div into a span or so the styling will break. You need to remember to adjust the styles as well. If you forget you'll end up with broken design in production

jkettmann commented on src/components/Header/Header.css:
```diff
@@ -0,0 +1,34 @@
+}
+
+.menu a:focus {
+  outline: none;
```
I think it's actually recommended to leave outline as is. This is for [accessibility](https://a11yproject.com/posts/never-remove-css-outlines/)

jkettmann commented on src/components/Header/Header.js:
```diff
@@ -0,0 +1,26 @@
+      <nav>
+        <ul className="menu">
+          <li>
+            <Link to="/search">Search</Link>
```
Style these with their own CSS classes as mentioned in the css file

jkettmann commented on src/components/Header/Header.js:
```diff
@@ -0,0 +1,26 @@
+        <Link to="/">Logo</Link>
+      </div>
+      <nav>
+        <ul className="menu">
```
Here as well. Rather use css-modules. `<ul className={styles.menu}>`

rotegras committed:
fixed code.
added css-in-js styled component.
added component basic structure.

jkettmann commented on src/App.css:
```diff
@@ -1,66 +1,22 @@
   flex-grow: 1;
 }
 
 h1 {
```
This should go to global.css as well

jkettmann commented on src/components/ResultsTable.js:
```diff
@@ -0,0 +1,20 @@
+import React from 'react';
+import Styled from 'styled-components';
+
+const SRTable = Styled.div`
```
What does SR stand for? You can call it `Wrapper` or `TableWrapper` for example

jkettmann commented on src/global.css:
```diff
@@ -0,0 +1,56 @@
+@import-normalize; /* bring in normalize.css styles */
```
You can use styled-components' [createGlobalStyle](https://styled-components.com/docs/api#createglobalstyle) instead of global.css

jkettmann commented on src/components/Table.js:
```diff
@@ -0,0 +1,19 @@
+import React from 'react';
+import Styled from 'styled-components';
+
+const STable = Styled.div`
```
Here as well `TableWrapper` or so. Or `StyledTable`. I would use abbreviations only in rare situations

jkettmann replied:
I saw that you use `StyledX` in other places. So naming them here as well like that would be more consistent

jkettmann reviewed (changes requested):
I left a few comments. One additional note: 

You're pretty fast but touching a lot of files here already. Usually, you would implement ticket by ticket. Here you're working on the Setup, Header, Hero, Subreddit Form, and Table ticket all at once.

This makes reviewing harder since the PRs get very big. In a real team this would also lead to conflicts with others since one person should be assigned one ticket in the In Progress column.

I'd suggest to stop working on features not related to this ticket. Finish the migration to styled-components and let's get this merged.

If you want to continue on the next feature already you can create a new branch from this branch and continue there.

rotegras committed:
fixed css consistency:
Fixed styled names.
Fonts loaded in index.html.
GloblalStyles replaces global.css
Normalize with package in GlobalStyles
